lib/opam/opam_prepare.ml

```diff
@@ -0,0 +1,174 @@
+open Core.Std
+open Core_extended.Std
+open Async.Std
+open Async_extra.Std
+
+exception Retrieval_failure of String.t
+
+let create_directory_name
+  : name:String.t -> semver:String.t -> String.t =
+  fun ~name ~semver ->
+    name ^ "." ^ semver
+
+let write_opam
+  : target_dir:String.t -> license:String.t
+  -> maintainer:String.t Option.t -> author:String.t Option.t
+  -> homepage:String.t Option.t -> bug_reports:String.t Option.t -> dev_repo:String.t
+  -> build_cmds:String.t List.t -> install_cmds:String.t List.t
+  -> remove_cmds:String.t List.t -> depends:String.t List.t
+  -> build_depends:String.t List.t -> (Unit.t, Exn.t) Deferred.Result.t =
+  fun ~target_dir ~license ~maintainer
+    ~author ~homepage ~bug_reports ~dev_repo ~build_cmds ~install_cmds
+    ~remove_cmds ~depends ~build_depends ->
+    Opam_make_opam.write_opam ~target_dir ~license ~maintainer
+    ~author ~homepage ~bug_reports ~dev_repo ~build_cmds ~install_cmds
+    ~remove_cmds ~depends ~build_depends ()
+
+let write_description
+  : opam_dir:String.t -> description_file:String.t -> (Unit.t, Exn.t) Deferred.Result.t =
+  fun ~opam_dir ~description_file ->
+    Reader.file_contents description_file
+    >>= fun body ->
+    Afin.Files.dump ~dir:opam_dir ~name:"descr" ~contents:body
+
+let make_url
+  : org:String.t -> name:String.t -> semver:String.t -> Uri.t =
+  fun ~org ~name ~semver ->
+    let path = "/" ^ org ^ "/" ^ name ^ "/archive/" ^ semver ^ ".tar.gz" in
+    Uri.make ~scheme:"https" ~host:"github.com" ~path ()
+
+let rec redirect
+  : logger:Log.t -> headers:Cohttp.Header.t -> (String.t, Exn.t) Deferred.Result.t =
+  fun ~logger ~headers ->
+    match Cohttp.Header.get headers "location" with
+    | Some loc ->
+      ignore @@ Log.info logger "Got redirected to %s" loc;
+      get_uri ~logger ~uri:(Uri.of_string loc)
+    | None ->
+      return @@ Error (Retrieval_failure "Redirect didn't have location")
+and get_uri
+  : logger:Log.t -> uri:Uri.t -> (String.t, Exn.t) Deferred.Result.t =
+  fun ~logger ~uri ->
+    ignore @@ Log.info logger "Retrieving tarbal from %s";
+    ignore @@ Vrt_common.Logging.flush logger;
+    Cohttp_async.Client.get uri
+    >>= function
+    | ({Cohttp.Response.status = `OK}, body) ->
+      Cohttp_async.Body.to_string body
+      >>= fun str_body ->
+      return @@ Ok str_body
+    | ({Cohttp.Response.status = `Found; headers}, body) ->
+      redirect ~logger ~headers
+    | (req, _) ->
+      return @@ Error (Retrieval_failure (Sexp.to_string @@ Cohttp.Response.sexp_of_t req;))
+
+let format_uri
+  : uri:Uri.t -> md5:String.t -> String.t =
+  fun ~uri ~md5 ->
+    "archive: \"" ^ (Uri.to_string uri) ^ "\"\n" ^
+    "checksum: \"" ^ md5 ^ "\""
```

Comment:
Ok, in the future we can submit a patch to them for it, should be simple. MD5 is very insecure, even for package signing these days.